lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,212 @@
+'use strict';
+
+const {
+  codes: { ERR_INVALID_ARG_VALUE },
+} = require('internal/errors');
+
+const {
+  ObjectEntries,
+  ArrayIsArray,
+  StringPrototypeToLocaleLowerCase,
+  StringPrototypeReplace,
+  Symbol,
+  SymbolIterator,
+  Array,
+} = primordials;
+
+const { validateObject } = require('internal/validators');
+
+const { isBoxedPrimitive } = require('internal/util/types');
+
+const { validateHeaderName, validateHeaderValue } = require('_http_outgoing');
+
+const { Buffer } = require('buffer');
+const console = require('console');
+
+const kHeadersList = Symbol('headers list');
+
+/**
+ * This algorithm is based off of
+ * https://www.tbray.org/ongoing/When/200x/2003/03/22/Binary
+ * It only operates on the even indexes of the array (the header names) by only
+ * iterating at most half the length of the input array. The search also
+ * assumes all entries are strings and uses String.prototype.localeCompare for
+ * comparison.
+ */
+function binarySearch(arr, val) {
+  let low = 0;
+  let high = arr.length / 2;
```

Comment:
```suggestion
  let high = MathFloor(arr.length / 2);
```
The source is in Java, so this does integer division. That's not the case in JS, so explicitly passing this through a `floor()`.